### PR TITLE
Handle Agg Functions with Literal Args When Used with a Union

### LINKED ISSUE #13305: PERCENTILEEST throws NPE in MultiStage Engine
we found in Pinot 1.1, PERCENTILEEST throws NPE in MultiStage Engine. Does anyone know if this is already fixed in master?

example query:
```
WITH sub AS (select deviceOS, count(1) as total_count from userAttributes group by 1 limit 100000),
subx AS (select deviceOS, count(1) as total_count from userAttributes group by 1 limit 100000),
union_table AS (select deviceOS, total_count from sub UNION ALL select deviceOS, total_count from subx)
select deviceOS, percentile(total_count, 50) from union_table group by deviceOS 
```

image

### PR DESCRIPTION
**Pull request by ankitsultana**
Attempts to fix #13305.

This is definitely not the most scientific fix (it's more like an empirical approach).

The issue was that aggregate node may not necessarily be on top of a project node. In our case, it was on top of a Union node, so this PR extends the aggregate exchange rule to handle that.

**Test Plan**

This query was failing before with the error below, but it works now:

```
SET useMultistageEngine=true;

with
  teamOne as (
  	select playerName, teamID, percentile(runs, 50) as sum_of_runs from baseballStats where teamID = 'SFN' group by playerName, teamID
  ),
  teamTwo as (
	select playerName, teamID, percentile(runs, 50) as sum_of_runs from baseballStats where teamID = 'CHN' group by playerName, teamID
  ),
  all as (
	select playerName, teamID, sum_of_runs from teamOne union all select playerName, teamID, sum_of_runs from teamTwo
  )
  select 
    /*+ aggOptions(is_skip_leaf_stage_group_by='true') */
    teamID, 
	-- sum(sum_of_runs) 
	percentile(sum_of_runs, 50) 
  from all
  group by teamID
```

Error:

```
Caused by: org.apache.pinot.spi.exception.BadQueryRequestException: Invalid aggregation function: percentile($1,$2); Reason: null
	at org.apache.pinot.core.query.aggregation.function.AggregationFunctionFactory.getAggregationFunction(AggregationFunctionFactory.java:487)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.getAggFunction(AggregateOperator.java:224)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.getAggFunctions(AggregateOperator.java:199)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.<init>(AggregateOperator.java:79)
org.apache.pinot.query.service.dispatch.QueryDispatcher.submit(QueryDispatcher.java:198)
org.apache.pinot.query.service.dispatch.QueryDispatcher.submitAndReduce(QueryDispatcher.java:95)
org.apache.pinot.broker.requesthandler.MultiStageBrokerRequestHandler.handleRequest(MultiStageBrokerRequestHandler.java:217)
org.apache.pinot.broker.requesthandler.BaseBrokerRequestHandler.handleRequest(BaseBrokerRequestHandler.java:133)
```

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 1]
```diff
@@ -220,7 +221,7 @@ private static PinotLogicalAggregate convertAggFromIntermediateInput(RelOptRuleC
       PinotLogicalExchange exchange, AggType aggType) {
     Aggregate aggRel = call.rel(0);
     RelNode input = PinotRuleUtils.unboxRel(aggRel.getInput());
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:223]: `unboxRel` is done within `findImmediateProjects()`, no need to unbox here

File: pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 2]
```diff
@@ -259,7 +260,7 @@ private static PinotLogicalAggregate convertAggFromIntermediateInput(RelOptRuleC
 
   private static List<AggregateCall> buildAggCalls(Aggregate aggRel, AggType aggType) {
     RelNode input = PinotRuleUtils.unboxRel(aggRel.getInput());
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:262]: Same here

File: pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 3]
```diff
@@ -330,4 +331,15 @@ private static AggregateCall buildAggCall(RelNode input, AggregateCall orgAggCal
         ImmutableList.of(), aggType.isInputIntermediateFormat() ? -1 : orgAggCall.filterArg, orgAggCall.distinctKeys,
         orgAggCall.collation, numGroups, input, null, null);
   }
+
+  @Nullable
+  private static List<RexNode> findImmediateProjects(RelNode relNode) {
+    relNode = PinotRuleUtils.unboxRel(relNode);
+    if (relNode instanceof Project) {
+      return ((Project) relNode).getProjects();
+    } else if (relNode instanceof Union) {
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:340]: Does this apply to other `RelNode` other than `Union`?

**Review by Jackie-Jiang — COMMENTED** [review]
Good catch!

File: pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java (hunk 3 above)
Comment (ankitsultana) [reply pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:340]: Yeah there could be more. For instance, if we have a Filter or even a Sort, we should ideally be able to hoist a literal up to the aggregate node.

I am a bit concerned if that could have edge-cases, that's why I kept the scope of the change in this PR small.

(digression follows)

Ideally, I think we should do this step in a separate optimization rule (run before agg exchange insert) which is run only for aggregate nodes.

Clubbing this with exchange insertion makes it a bit complicated to reason about.

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**